File: src/api/unified/statistics.cpp
// Copyright 2023 The EA Authors.
// part of Elastic AI Search
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#include <common/deprecated.hpp>
#include <fly/array.h>
#include <fly/statistics.h>
#include "symbol_manager.hpp"

fly_err fly_mean(fly_array *out, const fly_array in, const dim_t dim) {
    CHECK_ARRAYS(in);
    CALL(fly_mean, out, in, dim);
}

fly_err fly_mean_weighted(fly_array *out, const fly_array in,
                        const fly_array weights, const dim_t dim) {
    CHECK_ARRAYS(in, weights);
    CALL(fly_mean_weighted, out, in, weights, dim);
}

FLY_DEPRECATED_WARNINGS_OFF
fly_err fly_var(fly_array *out, const fly_array in, const bool isbiased,
              const dim_t dim) {
    CHECK_ARRAYS(in);
    CALL(fly_var, out, in, isbiased, dim);
}
FLY_DEPRECATED_WARNINGS_ON

fly_err fly_var_weighted(fly_array *out, const fly_array in, const fly_array weights,
                       const dim_t dim) {
    CHECK_ARRAYS(in, weights);
    CALL(fly_var_weighted, out, in, weights, dim);
}

fly_err fly_meanvar(fly_array *mean, fly_array *var, const fly_array in,
                  const fly_array weights, const fly_var_bias bias,
                  const dim_t dim) {
    CHECK_ARRAYS(in, weights);
    CALL(fly_meanvar, mean, var, in, weights, bias, dim);
}

FLY_DEPRECATED_WARNINGS_OFF
fly_err fly_stdev(fly_array *out, const fly_array in, const dim_t dim) {
    CHECK_ARRAYS(in);
    CALL(fly_stdev, out, in, dim);
}

fly_err fly_cov(fly_array *out, const fly_array X, const fly_array Y,
              const bool isbiased) {
    CHECK_ARRAYS(X, Y);
    CALL(fly_cov, out, X, Y, isbiased);
}
FLY_DEPRECATED_WARNINGS_ON

fly_err fly_median(fly_array *out, const fly_array in, const dim_t dim) {
    CHECK_ARRAYS(in);
    CALL(fly_median, out, in, dim);
}

fly_err fly_mean_all(double *real, double *imag, const fly_array in) {
    CHECK_ARRAYS(in);
    CALL(fly_mean_all, real, imag, in);
}

fly_err fly_mean_all_weighted(double *real, double *imag, const fly_array in,
                            const fly_array weights) {
    CHECK_ARRAYS(in, weights);
    CALL(fly_mean_all_weighted, real, imag, in, weights);
}

FLY_DEPRECATED_WARNINGS_OFF
fly_err fly_var_all(double *realVal, double *imagVal, const fly_array in,
                  const bool isbiased) {
    CHECK_ARRAYS(in);
    CALL(fly_var_all, realVal, imagVal, in, isbiased);
}
FLY_DEPRECATED_WARNINGS_ON

fly_err fly_var_all_weighted(double *realVal, double *imagVal, const fly_array in,
                           const fly_array weights) {
    CHECK_ARRAYS(in, weights);
    CALL(fly_var_all_weighted, realVal, imagVal, in, weights);
}

FLY_DEPRECATED_WARNINGS_OFF
fly_err fly_stdev_all(double *real, double *imag, const fly_array in) {
    CHECK_ARRAYS(in);
    CALL(fly_stdev_all, real, imag, in);
}
FLY_DEPRECATED_WARNINGS_ON

fly_err fly_median_all(double *realVal, double *imagVal, const fly_array in) {
    CHECK_ARRAYS(in);
    CALL(fly_median_all, realVal, imagVal, in);
}

fly_err fly_corrcoef(double *realVal, double *imagVal, const fly_array X,
                   const fly_array Y) {
    CHECK_ARRAYS(X, Y);
    CALL(fly_corrcoef, realVal, imagVal, X, Y);
}

fly_err fly_topk(fly_array *values, fly_array *indices, const fly_array in,
               const int k, const int dim, const fly_topk_function order) {
    CHECK_ARRAYS(in);
    CALL(fly_topk, values, indices, in, k, dim, order);
}

fly_err fly_var_v2(fly_array *out, const fly_array in, const fly_var_bias bias,
                 const dim_t dim) {
    CHECK_ARRAYS(in);
    CALL(fly_var_v2, out, in, bias, dim);
}

fly_err fly_var_all_v2(double *realVal, double *imagVal, const fly_array in,
                     const fly_var_bias bias) {
    CHECK_ARRAYS(in);
    CALL(fly_var_all_v2, realVal, imagVal, in, bias);
}

fly_err fly_cov_v2(fly_array *out, const fly_array X, const fly_array Y,
                 const fly_var_bias bias) {
    CHECK_ARRAYS(X, Y);
    CALL(fly_cov_v2, out, X, Y, bias);
}

fly_err fly_stdev_v2(fly_array *out, const fly_array in, const fly_var_bias bias,
                   const dim_t dim) {
    CHECK_ARRAYS(in);
    CALL(fly_stdev_v2, out, in, bias, dim);
}

fly_err fly_stdev_all_v2(double *real, double *imag, const fly_array in,
                       const fly_var_bias bias) {
    CHECK_ARRAYS(in);
    CALL(fly_stdev_all_v2, real, imag, in, bias);
}

File: python/flare/statistics.py
##########################################################################
# Copyright 2023 The EA Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

"""
Statistical algorithms (mean, var, stdev, etc).
"""

from .library import *
from .array import *

def mean(a, weights=None, dim=None):
    """
    Calculate mean along a given dimension.

    Parameters
    ----------
    a: fly.Array
        The input array.

    weights: optional: fly.Array. default: None.
        Array to calculate the weighted mean. Must match size of the
        input array.

    dim: optional: int. default: None.
        The dimension for which to obtain the mean from input data.

    Returns
    -------
    output: fly.Array
        Array containing the mean of the input array along a given
        dimension.
    """
    if dim is not None:
        out = Array()

        if weights is None:
            safe_call(backend.get().fly_mean(c_pointer(out.arr), a.arr, c_int_t(dim)))
        else:
            safe_call(backend.get().fly_mean_weighted(c_pointer(out.arr), a.arr, weights.arr, c_int_t(dim)))

        return out
    else:
        real = c_double_t(0)
        imag = c_double_t(0)

        if weights is None:
            safe_call(backend.get().fly_mean_all(c_pointer(real), c_pointer(imag), a.arr))
        else:
            safe_call(backend.get().fly_mean_all_weighted(c_pointer(real), c_pointer(imag), a.arr, weights.arr))

        real = real.value
        imag = imag.value

        return real if imag == 0 else real + imag * 1j

def var(a, bias=VARIANCE.DEFAULT, weights=None, dim=None):
    """
    Calculate variance along a given dimension.

    Parameters
    ----------
    a: fly.Array
        The input array.

    bias: optional: fly.VARIANCE. default: DEFAULT.
        population variance(VARIANCE.POPULATION) or sample variance(VARIANCE.SAMPLE).
        This is ignored if weights are provided.

    weights: optional: fly.Array. default: None.
        Array to calculate for the weighted mean. Must match size of
        the input array.

    dim: optional: int. default: None.
        The dimension for which to obtain the variance from input data.

    Returns
    -------
    output: fly.Array
        Array containing the variance of the input array along a given
        dimension.
    """
    if dim is not None:
        out = Array()

        if weights is None:
            safe_call(backend.get().fly_var_v2(c_pointer(out.arr), a.arr, bias.value, c_int_t(dim)))
        else:
            safe_call(backend.get().fly_var_weighted(c_pointer(out.arr), a.arr, weights.arr, c_int_t(dim)))

        return out
    else:
        real = c_double_t(0)
        imag = c_double_t(0)

        if weights is None:
            safe_call(backend.get().fly_var_all_v2(c_pointer(real), c_pointer(imag), a.arr, bias.value))
        else:
            safe_call(backend.get().fly_var_all_weighted(c_pointer(real), c_pointer(imag), a.arr, weights.arr))

        real = real.value
        imag = imag.value

        return real if imag == 0 else real + imag * 1j

def meanvar(a, weights=None, bias=VARIANCE.DEFAULT, dim=-1):
    """
    Calculate mean and variance along a given dimension.

    Parameters
    ----------
    a: fly.Array
        The input array.

    weights: optional: fly.Array. default: None.
        Array to calculate for the weighted mean. Must match size of
        the input array.

    bias: optional: fly.VARIANCE. default: DEFAULT.
        population variance(VARIANCE.POPULATION) or
        sample variance(VARIANCE.SAMPLE).

    dim: optional: int. default: -1.
        The dimension for which to obtain the variance from input data.

    Returns
    -------
    mean: fly.Array
        Array containing the mean of the input array along a given
        dimension.
    variance: fly.Array
        Array containing the variance of the input array along a given
        dimension.
    """

    mean_out = Array()
    var_out  = Array()

    if weights is None:
        weights  = Array()

    safe_call(backend.get().fly_meanvar(c_pointer(mean_out.arr), c_pointer(var_out.arr),
                                       a.arr, weights.arr, bias.value, c_int_t(dim)))

    return mean_out, var_out


def stdev(a, bias=VARIANCE.DEFAULT, dim=None):
    """
    Calculate standard deviation along a given dimension.

    Parameters
    ----------
    a: fly.Array
        The input array.

    bias: optional: fly.VARIANCE. default: DEFAULT.
        population variance(VARIANCE.POPULATION) or sample variance(VARIANCE.SAMPLE).
        This is ignored if weights are provided.

    dim: optional: int. default: None.
        The dimension for which to obtain the standard deviation from
        input data.

    Returns
    -------
    output: fly.Array
        Array containing the standard deviation of the input array
        along a given dimension.
    """
    if dim is not None:
        out = Array()
        safe_call(backend.get().fly_stdev_v2(c_pointer(out.arr), a.arr, bias.value,
                                            c_int_t(dim)))
        return out
    else:
        real = c_double_t(0)
        imag = c_double_t(0)
        safe_call(backend.get().fly_stdev_all_v2(c_pointer(real), c_pointer(imag), a.arr,
                                                bias.value))
        real = real.value
        imag = imag.value
        return real if imag == 0 else real + imag * 1j

def cov(a, b, bias=VARIANCE.DEFAULT):
    """
    Calculate covariance along a given dimension.

    Parameters
    ----------
    a: fly.Array
        Input array.

    b: fly.Array
        Input array.

    bias: optional: fly.VARIANCE. default: DEFAULT.
        population variance(VARIANCE.POPULATION) or sample variance(VARIANCE.SAMPLE).

    Returns
    -------
    output: fly.Array
        Array containing the covariance of the input array along a given dimension.
    """
    out = Array()
    safe_call(backend.get().fly_cov_v2(c_pointer(out.arr), a.arr, b.arr, bias.value))
    return out

def median(a, dim=None):
    """
    Calculate median along a given dimension.

    Parameters
    ----------
    a: fly.Array
        The input array.

    dim: optional: int. default: None.
        The dimension for which to obtain the median from input data.

    Returns
    -------
    output: fly.Array
        Array containing the median of the input array along a
        given dimension.
    """
    if dim is not None:
        out = Array()
        safe_call(backend.get().fly_median(c_pointer(out.arr), a.arr, c_int_t(dim)))
        return out
    else:
        real = c_double_t(0)
        imag = c_double_t(0)
        safe_call(backend.get().fly_median_all(c_pointer(real), c_pointer(imag), a.arr))
        real = real.value
        imag = imag.value
        return real if imag == 0 else real + imag * 1j

def corrcoef(x, y):
    """
    Calculate the correlation coefficient of the input arrays.

    Parameters
    ----------
    x: fly.Array
        The first input array.

    y: fly.Array
        The second input array.

    Returns
    -------
    output: fly.Array
        Array containing the correlation coefficient of the input arrays.
    """
    real = c_double_t(0)
    imag = c_double_t(0)
    safe_call(backend.get().fly_corrcoef(c_pointer(real), c_pointer(imag), x.arr, y.arr))
    real = real.value
    imag = imag.value
    return real if imag == 0 else real + imag * 1j

def topk(data, k, dim=0, order=TOPK.DEFAULT):
    """
    Return top k elements along a single dimension.

    Parameters
    ----------

    data: fly.Array
          Input array to return k elements from.

    k: scalar. default: 0
       The number of elements to return from input array.

    dim: optional: scalar. default: 0
         The dimension along which the top k elements are
         extracted. Note: at the moment, topk() only supports the
         extraction of values along the first dimension.

    order: optional: fly.TOPK. default: fly.TOPK.DEFAULT
           The ordering of k extracted elements. Defaults to top k max values.

    Returns
    -------

    values: fly.Array
            Top k elements from input array.
    indices: fly.Array
             Corresponding index array to top k elements.
    """

    values = Array()
    indices = Array()

    safe_call(backend.get().fly_topk(c_pointer(values.arr), c_pointer(indices.arr), data.arr, k, c_int_t(dim), order.value))

    return values,indices
